src/snap_management.py

```diff
@@ -0,0 +1,180 @@
+
+        Returns:
+            A dict of SnapArguments keyed by their snap name
+        """
+        manifest_raw = self._get_manifest()
+        return self._parse_manifest(manifest_raw)
+
+    def _get_manifest(self) -> List[Dict]:
+        """Loads the snap manifest file, returning the manifest for this architecture.
+
+        Raises:
+            SnapManifestError: when the manifest file cannot be loaded or parsed.
+
+        Returns:
+            A list of dicts describing snaps
+        """
+        if not self._manifest_path.exists():
+            raise SnapManifestError(f"Failed to find file={self._manifest_path}")
+        try:
+            with self._manifest_path.open() as f:
+                manifest = yaml.safe_load(f)
+        except yaml.YAMLError as e:
+            log.error("Failed to load manifest file=%s, %s", self._manifest_path, e)
+            raise SnapManifestError(f"Failed to load manifest file={self._manifest_path}")
+
+        manifest_this_arch = manifest.get(self.arch)
+        if not (isinstance(manifest, dict) and manifest_this_arch):
+            log.warning("Failed to find revision for arch=%s", self.arch)
+            raise SnapManifestError(f"Failed to find revision for arch={self.arch}")
+
+        return manifest_this_arch
+
+    def _parse_manifest(self, manifest: List[Dict]) -> Dict[str, SnapArgument]:
+        """Parses a manifest defined by a list of snap dicts, rendering them to SnapArguments."""
+        try:
+            manifest_parsed = {
+                arg["name"]: parse_obj_as(SnapArgument, arg) for arg in manifest  # pyright: ignore
+            }
+        except (ValidationError, KeyError) as e:
+            log.warning("Failed to validate args=%s (%s)", self.arch, e)
+            raise SnapManifestError("Failed to validate snap args")
+
+        return manifest_parsed
+
+    def get_snap_arguments(self, name: str):
+        """Returns a SnapArgument for the given snap name."""
+        return self._manifest[name]
+
+
+def install_snap(snap_arguments: SnapArgument):
+    """Install the snap defined by a SnapArgument."""
+    cache = snap_lib.SnapCache()
+    snap = cache[snap_arguments.name]
+    if isinstance(snap_arguments, SnapFileArgument) and snap.revision != "x1":
+        snap_lib.install_local(**snap_arguments.dict(exclude_none=True))
+    elif isinstance(snap_arguments, SnapStoreArgument):
+        log.info(
+            "Ensuring %s snap is installed with channel=%s, revision=%s",
+            snap_arguments.name,
+            snap_arguments.channel,
+            snap_arguments.revision,
+        )
+        snap.ensure(**snap_arguments.dict(exclude_none=True))
+        # TODO: should hold be an argument in the yaml?
+        if snap_arguments.revision:
+            log.info(
+                "Setting snap refresh for %s to hold=forever because revision is defined",
+                snap_arguments.name,
+            )
+            snap.hold()
```

Comment:
If we use `snap.hold()`, why do we need snap_management and the new yaml?

Reply:
(I'll also clarify this in the PR)

The goals here are to:
1. ensure the snap version installed is "held".  That way someone pushing a new `latest/stable` cannot break someone's existing deployment of the charm
2. define what snap version this specific charm revision uses.  That way it is predictable and reproducible later (if someone deploys charm rev100 next year, they should get the same snap as if someone deployed it today)

doing `snap.hold()` achieves (1), whereas the snap_management/new yaml achieves (2).  Although tbh, snap_management/new yaml might be heavier than we need to achieve (2)

Reply:
after talking with @sed-i I've added a competing PR #131 taking a simpler, more hard-coded approach